models/resnet/resnet_impl.hpp

```diff
@@ -0,0 +1,103 @@
+/**
+ * @file resnet_impl.hpp
+ * @author Aakash Kaushik
+ *
+ * Implementation of ResNet using mlpack.
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MODELS_MODELS_RESNET_RESNET_IMPL_HPP
+#define MODELS_MODELS_RESNET_RESNET_IMPL_HPP
+
+#include "resnet.hpp"
+
+namespace mlpack {
+namespace models {
+
+template<typename OutputLayerType, typename InitializationRuleType,
+    size_t ResNetVersion>
+ResNet<OutputLayerType, InitializationRuleType, ResNetVersion>::ResNet() :
+    inputChannel(0),
+    inputWidth(0),
+    inputHeight(0),
+    numClasses(0),
+{
+  // Nothing to do here.
+}
+
+template<typename OutputLayerType, typename InitializationRuleType,
+    size_t ResNetVersion>
+ResNet<OutputLayerType, InitializationRuleType, ResNetVersion>::ResNet(
+    const bool includeTop,
+    const bool preTrained,
+    const size_t inputChannel,
+    const size_t inputWidth,
+    const size_t inputHeight,
+    const size_t numClasses) :
+    ReNet<OutputLayerType, InitializationRule,  ResNetVersion>(
+        includeTop,
+        preTrained,
+        std::tuple<size_t, size_t, size_t>(
+        inputChannel,
+        inputWidth,
+        inputHeight),
+        numClasses)
+{
+  // Nothing to do here.
+}
+
+template<typename OutputLayerType, typename InitializationRuleType,
+    size_t ResNetVersion>
+ResNet<OutputLayerType, InitializationRuleType, ResNetVersion>::ResNet(
+    const bool includeTop,
+    const bool preTrained,
+    std::tuple<size_t, size_t, size_t> inputShape,
+    const size_t numClasses,
+    ) :
+    inputChannel(std::get<0>(inputShape)),
+    inputWidth(std::get<1>(inputShape)),
+    inputHeight(std::get<2>(inputShape)),
+    numClasses(numClasses)
+{
+
+  resNet.Add(ann::Convolution<>(3, 64, 7, 7, 2, 2, 3, 3, inputWidth, inputHeight));
+  
+  // Updating input dimesntions.
+  inputwidth = ConvOutSize(inputWidth, kernelWidth, strideWidth, padW);
+  inputHeight = ConvOutSize(inputHeight, kernelHeight, strideHeight, padH);
+  
+  resNet.Add(ann::batchNorm<>(64));
+  resNet.Add(ann::ReLULayer<>);
+  
+  resNet.Add(ann::Padding<>(1, 1, 1, 1)) 
+  resNet.Add(ann::MaxPooling<>(3, 3, 2, 2));
+
+  if (ResNetVersion == 18)
+  {
+    size_t numBlockArray = [2 ,2, 2, 2]  
+  }
+
+  MakeLayer("basicblock", 64, numBlockArray[0]);
+  MakeLayer("basicblock", 128, numBlockArray[1]);
+  MakeLayer("basicblock", 256, numBlockArray[2]);
+  MakeLayer("basicblock", 512, numBlockArray[3]);
+
+  // What would be the Pytroch equivalent of nn.AdaptiveAvgPool2d((1, 1))
```

Comment:
And also a doubt in this.

Reply:
Hey @zoq, @kartikdutt18 i had these two doubts.

Reply:
I sorry I forget to provide an answer for that one, there is no `AdaptivePooling` in mlpack, but as long as the input size doesn't change you can use `AveragePooling`, but we have to make sure we don't pass different image sizes to the model.

Reply:
I think Adaptive Pooling is Adaptive Max Pooling or Adaptive Mean Pooling. Link : https://github.com/mlpack/mlpack/blob/master/src/mlpack/methods/ann/layer/adaptive_max_pooling_impl.hpp

Reply:
Hey @zoq as @kartikdutt18 pointed out we do have adaptive max and mean pools and i have used that one only but for now i have seg faults in my model and no good way to debug it gdb throws out bunch of errors and i am not too experienced with it, so how should i go forward now ? I was also thinking to take a look at implementing something in the ffn class in mlpack to print model summary.

Reply:
Try building model block by block and call forward propagation at after each block to find out at which block the model segfaults. Use that info to determine cause.

Reply:
> I think Adaptive Pooling is Adaptive Max Pooling or Adaptive Mean Pooling. Link : https://github.com/mlpack/mlpack/blob/master/src/mlpack/methods/ann/layer/adaptive_max_pooling_impl.hpp

Ohh, I missed that one, `AdaptiveMeanPooling` would be the equivalent then.

Reply:
> Try building model block by block and call forward propagation, after adding each block to find out at which block the model segfaults. Use that info to determine cause.

Wouldn't building a model summary like thing be more helpful, that would give me each layer and their params. and would kill 2 things with one. Also i don't feel like that maybe too much work for just debugging and would at the end be deleted.

Reply:
Try one thing first, add cout statements in blocks and print block name and input and output sizes. Should be sufficient to debug.

Reply:
Yup that should be better. thanks.

Reply:
Prints out a simple layer with in and out size and that was incorrect made that work but still have a seg fault and now i believe that it is mostly due to my lack of undertanding of how padding or certain params work in mlpack and that is messing up with the dimensions. btw also slight congo we were able to replicate the resnet18 architecture and so the other ones should be easy once i can figure the seg fault.